crates/egui/src/memory.rs

```diff
@@ -80,8 +81,9 @@ pub struct Memory {
 
     /// Which popup-window is open (if any)?
     /// Could be a combo box, color picker, menu etc.
+    /// The bool is used to detect if a popup stopped showing before calling close_popup
     #[cfg_attr(feature = "persistence", serde(skip))]
-    popup: Option<Id>,
+    popup: Option<(Id, std::sync::Arc<AtomicBool>)>,
```

Comment:
I'm using an `AtomicBool` here because I didn't want to change `is_popup_open` to take `&mut self`.

An alternative is to add another function, something like `retain_popup(&mut self)` that explicitly sets a `retained` bool to true. But would will be a breaking change and existing apps will need to call it or they popups will last for one frame.

I'm open to suggestions.

Reply:
it's not at all obvious from reading this what the bool is for though., or why it is behind an `Arc`.

It seems to answer the question "Was `is_popup_open` called for this this frame?", i.e. "Does anyone care that this is open?".

I suggest making a `struct` with a name field so the motivation behind this can be documented.

We should also document that if `is_popup_open` is not called during a frame, the popup auto-closes.
Since this in fact makes it change state, I think making it take a `&mut` makes sense.

So I suggest

```
struct OpenPopup {
    id: Id,
    shown_this_frame: bool,
}

…

popup: Option<OpenPopup>,
```